text/0000-bind-helper.md

```diff
@@ -0,0 +1,113 @@
+- Start Date: 2019-03-20
+- Relevant Team(s): Ember.js, Learning
+- RFC PR: https://github.com/emberjs/rfcs/pull/470
+- Tracking: (leave this empty)
+
+# `{{fn}}` Helper
+
+## Summary
+
+This RFC introduces a new helper, `{{fn}}`, to allow clear argument passing for functions in templates.
+
+## Motivation
+
+The current `action` helper has a lot of implicit and confusing behavior that is different than the Octane and post Octane programming model.
+
+To understand the complexity of `action` there are many complex behaviors including:
+
+1. Argument partial application (currying)
+2. `this` context binding
+3. `send` checks for Component and Controllers
+
+At build time the `action` helper currently is passed through an AST transform to explicitly bind the `this` to be deterministic at runtime. This is a private API where the outputted Glimmer is not a 1-1 to the template. Also, the `action` helper is confused and has overlap with the `action` modifier which has similar but slightly different behavior.
+
+Instead of this confusing and overloaded behavior, a new `fn` helper would be introduced to do partial application (with no need for build time private APIs), and context binding will be done instead using the `@action` decorator in classes.
+
+## Detailed design
+
+The `fn` helper will take in a function and then the set of arguments that will be partially applied to the function.
+
+Here are some examples of the `fn` helper and the equivalent JS:
+
+### Simple Case On Argument Curry
+
+```hbs
+{{fn this.log 1}}
+```
+
+```js
+return function() {
+  this.log.call(this, 1);
+}
+```
+
+### Multiple Argument Partial Application
+
+```hbs
+{{fn this.add 1 2}}
+```
+
+```js
+return function() {
+  this.add.call(this, 1, 2);
+}
+```
```

Comment:
I got a bit confused by looking at these examples, as they make it look like you have to pass *all* expected arguments, rather than partially applying them. Wouldn't the correct JS equivalent of `{{fn this.add 1 2}}` look more like this?

```js
return function(...args) {
  this.add.call(this, 1, 2, ...args);
}
```

Reply:
Yeah actually that’s totally correct, we will update them!

Reply:
@pzuraq it looks like this change didn't make it into the final version of the RFC text?